File: python/0017电话号码的字母组合.py

d2l_map=[
    "",
    "",
    "abc",
    "def",
    "ghi",
    "jkl",
    "mno",
    "pqrs",
    "tuv",
    "wxyz",]



def letterCombinations(digits: str) -> list:
    res=[]
    if(not len(digits)):
        return res
        
    def _back_tracking(temp:str,digits:str,s=0):
        if(s==len(digits)):
            res.append(temp)
            return
        
        for ch in d2l_map[int(digits[s])]:
            _back_tracking(temp+ch,digits,s+1)
            
    _back_tracking("",digits,0)
    return res

print(letterCombinations("23"))

File: cpp/0017电话号码的字母组合.cpp
#include<vector>
#include<iostream>
#include<vector>

using namespace std;

const string d2lmap[10] = {
    "",
    "",
    "abc",
    "def",
    "ghi",
    "jkl",
    "mno",
    "pqrs",
    "tuv",
    "wxyz",
};

vector<string> res;
string temp = "";

//建立回溯树，明确深度和广度，深度用递归遍历，广度用循环遍历
void backTracking(int s, vector<int> digits)
{
    if(s==digits.size())
    {
        res.push_back(temp);
        return;
    }

    for(char ch :d2lmap[digits[s]])
    {
        temp.push_back(ch);
        backTracking(s + 1, digits);
        temp.pop_back();
    }
    
}

vector<string> letterCombinations(string digits)
{

    if(digits.size())
    {
        int num = stoi(digits);
        int len = digits.size();
        vector<int> nums(len);
        do
        {
            nums[len - 1] = num % 10;
            
            num /= 10;
        } while (--len);

        for (auto i : nums)
        {
            cout << i;
        }

        backTracking(0, nums);
    }

    return res;
}


int main()
{
    letterCombinations("123");
}